File: treasure.pl
#!/path/to/bin/perl
# https://.../treasure.pl
# ?page=7013432&team=boys
# ?page=7013432&answer=hill&attempt=2&team=boys
# ?page=7013432&reveal=true&team=boys
# For Admin:
# ?admin=true&game=jer_heb
# ?game=jer_heb&team=Team1&showMap=true
use strict;
use CGI;
use CGI::Carp qw(fatalsToBrowser);
use Encode; #use utf8;
use Time::HiRes;
$| = 1; # unbuffered output
my $q=new CGI;
print $q->header(-charset=>'utf-8');
(my $SEC, my $MIN, my $HOUR, my $DAY,my $MONTH,my $YEAR) =
             (localtime)[0,1,2,3,4,5]; $MONTH+=1; $YEAR+=1900;
my $now = sprintf("%04d%02d%02d_%02d%02d%02d",$YEAR,$MONTH,$DAY,$HOUR,$MIN,$SEC);
#print $now;
my $path = '/exlibris/aleph/a22_1/aleph/cgi/treasure/';
my $game = $q->param('game');
my $team = $q->param('team');

# load conf:
my $conf_file = $path.'data/'.$game.'.conf';
my $conf_text = slurp($conf_file);
#print $conf_file , $conf_text;
my %conf;
while ($conf_text=~m/(.*?):\n(.*?)\n/gs) {
  #print $1,'=',$2,"\n";
  $conf{$1} = $2;
}

##### first page
if (!$team) {
  my $out_html = translate_html(slurp($path . 'html/first.htm'));
  $out_html=~s/{game}/$game/g;
  $out_html=~s/{game}/$game/g;
  output_html($out_html);
  exit;
}

##### admin page
if ($team eq 'admin') {
  foreach my $file ( glob( $path . 'data/team*' ) ) {
    print "$file <br/>\n <pre>";
    print slurp($file),"</pre>";
    print "___________________<br/><br/><br/><br/>\n";
  }
  exit;
}

my $team_file = $path.'data/team_'.$game.'_'.$team.'.data';
if ($q->param('showMap')) {  showMap($team_file); }

open( OUT_F, ">>$team_file") or die "Cannot open $team_file as output\n" ;
my $team_text = '';
my $current_points = 0;
my $start_time;
my $elapsed_time = 0;
  $team_text = slurp($team_file);
  if ($team_text!~/start timestamp/) {
    my $start = time();
    print OUT_F "start timestamp:$start\ntime:$now\n";
  }
  if ($team_text=~/.*points:(\d*)/s) {
    $current_points = $1;
  }
  if ($team_text=~/start timestamp:(\d*)/s) {
    $start_time = $1;
    $elapsed_time = wdhms(time() - $start_time);
  } else {
    $elapsed_time = wdhms(1); # for the first page: 1 sec
  }

my $page_param = $q->param('page');
my $page_no = substr($page_param,1,2);
my $the_question = $conf{$page_no.'_q'} ; #print "ooo  $page_no $the_question";

my $answer = $q->param('answer');
my $attempt = $q->param('attempt');
if (!$attempt) { $attempt = 1 };
##### question page:
if ($page_param && !$answer) {
  my $out_html = translate_html(slurp($path.'html/question.htm'));
  $out_html=~s/{page}/$page_param/g;
  $out_html=~s/{attempt}/$attempt/g;
  if ($conf{$page_no.'_q_img'}) {
    $the_question.='<br/><img style="width:100%; max-width:800px" src="' . $conf{$page_no.'_q_img'} . '"./>';
  }
  $out_html=~s/{question}/$the_question/g;
  output_html($out_html);
  print OUT_F "page: $page_no attempt:$attempt\n";
  exit;
}

##### post an answer:
# ?page=7013432&answer=hill&attempt=2&team=boys
if ($page_param && $answer) {

  # log geo location
  if ( $q->param('lat')  ){
    print OUT_F "GEO: lat " . $q->param('lat') . " lon " . $q->param('lon') . "\n";
  }

  my $correct_answer = $conf{$page_no.'_a'} ; print "<!-- ooo  $page_no $correct_answer -->";
  my $added_points = 12 - (2 * $attempt);
  if ($answer=~/$correct_answer/i) {
    $current_points += $added_points;
    my $out_html = translate_html(slurp($path.'html/correct.htm'));
    my $next_page = hidden_page_num($page_no+1);
    $out_html=~s/{page}/$next_page/g;
    $out_html=~s/{added_points}/$added_points/g;
    my $after_reply = $conf{$page_no.'_after_reply'};
    $out_html=~s/{after_reply}/$after_reply/g;
    output_html($out_html);
    print OUT_F "CORRECT page: $page_no attempt:$attempt correct_answer:$correct_answer . answer:$answer elapsed:$elapsed_time\n";
    print OUT_F "points:$current_points\n";
    exit;
  } else {
    print OUT_F "WRONG page: $page_no attempt:$attempt correct_answer:$correct_answer . answer:$answer elapsed:$elapsed_time\n";

    $attempt++;
    if ($attempt > 3) {
      ##### after too many wrong asnwers - show the answer
      my $out_html = translate_html(slurp($path.'html/after_many_mistakes.htm'));
      my $next_page = hidden_page_num($page_no+1);
      $out_html=~s/{page}/$next_page/g;
      my $after_reply = $conf{$page_no.'_after_reply'};
      $out_html=~s/{after_reply}/$after_reply/g;
      $out_html=~s/{answer}/$correct_answer/g;
      output_html($out_html);
      print OUT_F "MANY_WRONG page: $page_no attempt:$attempt correct_answer:$correct_answer . answer:$answer elapsed:$elapsed_time\n";
      print OUT_F "points:$current_points\n";
      exit;
    }

    my $out_html = translate_html(slurp($path.'html/wrong.htm'));
    $out_html=~s/{page}/$page_param/g;
    $out_html=~s/{attempt}/$attempt/g;
    if ($conf{$page_no.'_q_img'}) {
      $the_question.='<br/><img style="width:100%; max-width:800px;" src="' . $conf{$page_no.'_q_img'} . '"./>';
    }
    $out_html=~s/{question}/$the_question/g;
    output_html($out_html);
    exit;
  }
}

my $log_file = $path."logs/main.log";
open( OUT_F, ">>$log_file") or die "Cannot open $log_file as output\n" ;
print OUT_F "ooo $now\n\n";

print "error...\n";

exit;
###############################################################################
my $REMOTE_ADDR = $ENV{REMOTE_ADDR}; print " $REMOTE_ADDR \n";
print "<br><br>\n";
foreach my $key (sort keys(%ENV) ) { print "$key = $ENV{$key}<br>\n"; }
my @names = $q->param; my $c = 0;
foreach my $name (sort @names) { $c++; print "<br> $name ->", $q->param($name),"\n"; }
print "<br><br>\n";


###############################################################################
sub translate_html {
    my $html = shift;
    foreach my $key (keys %conf) {
        $html=~s/{$key}/$conf{$key}/g;
    }
    $html=~s/{game}/$game/g;
    $html=~s/{team}/$team/g;
    $html=~s/{current_points}/$current_points/g;
    $html=~s/{elapsed_time}/$elapsed_time/g;
    return $html;
}

###############################################################################
sub output_html{
    my $html = shift;
    print $html;
}

###############################################################################
sub hidden_page_num{
    my $num = shift;
    return '7' . sprintf("%02d", $num) . '3432';
}

###############################################################################
sub slurp {
    my $file = shift;
    open my $fh, '<', $file or die "can not find $file";
    local $/ = undef;
    my $cont = <$fh>;
    close $fh;
    return $cont;
}

###############################################################################
# from http://www.perlmonks.org/?node_id=110550
sub wdhms {
    my( $weeks, $days, $hours, $minutes, $seconds, $sign, $res ) = qw/+0 0 0 0 0/;

    $seconds = shift;
    $sign    = $seconds == abs $seconds ? '' : '-';
    $seconds = abs $seconds;

    ($seconds, $minutes) = ($seconds % 60, int($seconds / 60)) if $seconds;
    ($minutes, $hours  ) = ($minutes % 60, int($minutes / 60)) if $minutes;
    ($hours,   $days   ) = ($hours   % 24, int($hours   / 24)) if $hours;
    ($days,    $weeks  ) = ($days    %  7, int($days    /  7)) if $days;

    $res = sprintf '%02d',     $seconds;
    $res = sprintf "%02d:$res", $minutes if $minutes or $hours or $days or $weeks;
    $res = sprintf "%02d:$res", $hours   if             $hours or $days or $weeks;
    if ($days > 0) {
      $res = sprintf "%dd $res", $days    if                       $days or $weeks;
      $res = sprintf "%dw$res", $weeks   if                                $weeks;
    }

    return "$sign$res";
}

sub showMap {
    my $file = shift;
    $team_text = slurp($team_file);
    # GEO: lat 31.7493453 lon 35.1863789
    my $values = '';
    while ($team_text=~m/GEO: lat (.*?) lon (.*?)\n/gs) {
        $values .= ' ['.$1.', '.$2.', ""],'
    }
    my $html_text = slurp($path . 'html/showMap.htm');
    $html_text=~s/{values}/$values/;
    print $html_text;
    exit;
}

File: treasure.php
<?php

header('Content-Type: text/html; charset=UTF-8');

$path = __DIR__ . DIRECTORY_SEPARATOR;
$game = isset($_REQUEST['game']) ? (string) $_REQUEST['game'] : 'jer_heb';
if ($game === '') {
    $game = 'jer_heb';
}
$team = isset($_REQUEST['team']) ? (string) $_REQUEST['team'] : '';
$conf = loadConfig($path . 'data' . DIRECTORY_SEPARATOR . $game . '.conf');
$now = date('Ymd_His');
$currentPoints = 0;
$elapsedTime = 0;

if ($team === '') {
    $output = translateHtml(readFileContents($path . 'html' . DIRECTORY_SEPARATOR . 'first.htm'));
    outputHtml($output);
    exit;
}

if ($team === 'admin') {
    showAdmin($path, $game, $conf);
    exit;
}

$teamFile = $path . 'data' . DIRECTORY_SEPARATOR . 'team_' . $game . '_' . $team . '.data';

if (isset($_REQUEST['showMap'])) {
    showMap($teamFile, $path);
}

if (isset($_REQUEST['showLog'])) {
    showLog($teamFile);
}

$pageParam = isset($_REQUEST['page']) ? (string) $_REQUEST['page'] : '';
$pageNo = substr($pageParam, 1, 2);
$question = isset($conf[$pageNo . '_q']) ? $conf[$pageNo . '_q'] : '';
$answer = isset($_REQUEST['answer']) ? (string) $_REQUEST['answer'] : '';
$attempt = isset($_REQUEST['attempt']) ? (string) $_REQUEST['attempt'] : '';

if ($pageNo === '01' && $answer === '' && file_exists($teamFile)) {
    echo "Team name already taken. Please pick a new name\n";
    exit;
}

$teamText = file_exists($teamFile) ? readFileContents($teamFile) : '';
$teamHandle = fopen($teamFile, 'ab');
if ($teamHandle === false) {
    http_response_code(500);
    exit('Cannot open team file for output');
}

if (strpos($teamText, 'start timestamp') === false) {
    fwrite($teamHandle, 'start timestamp:' . time() . "\ntime:" . $now . "\n");
}

if (preg_match('/.*points:(\d*)/s', $teamText, $matches)) {
    $currentPoints = (int) $matches[1];
}
$elapsedTime = calcElapsed($teamText, $conf);

if ($attempt === '' || $attempt === '0') {
    $attempt = '1';
}

if ($pageParam !== '' && $answer === '') {
    if (isset($conf[$pageNo . '_q_img']) && $conf[$pageNo . '_q_img'] !== '') {
        $question .= '<br/><img style="width:100%; max-width:800px" src="' . $conf[$pageNo . '_q_img'] . '"./>';
    }
    $output = translateHtml(readFileContents($path . 'html' . DIRECTORY_SEPARATOR . 'question.htm'));
    $output = str_replace(
        array('{page}', '{attempt}', '{question}'),
        array($pageParam, $attempt, $question),
        $output
    );
    outputHtml($output);
    fwrite($teamHandle, 'page: ' . $pageNo . ' attempt:' . $attempt . "\n");
    fclose($teamHandle);
    exit;
}

if ($pageParam !== '' && $answer !== '') {
    if (isset($_REQUEST['lat']) && (string) $_REQUEST['lat'] !== '') {
        $longitude = isset($_REQUEST['lon']) ? $_REQUEST['lon'] : '';
        fwrite($teamHandle, 'GEO: lat ' . $_REQUEST['lat'] . ' lon ' . $longitude . "\n");
        fwrite($teamHandle, 'log time: ' . $now . "\n");
    }

    $correctAnswer = isset($conf[$pageNo . '_a']) ? $conf[$pageNo . '_a'] : '';
    echo '<!-- ooo ' . $pageNo . ' ' . $correctAnswer . ' -->';
    $addedPoints = 12 - (2 * (int) $attempt);

    if (preg_match('/' . $correctAnswer . '/iu', $answer) === 1) {
        $currentPoints += $addedPoints;
        $output = translateHtml(readFileContents($path . 'html' . DIRECTORY_SEPARATOR . 'correct.htm'));
        $output = str_replace(
            array('{page}', '{added_points}', '{after_reply}'),
            array(hiddenPageNum((int) $pageNo + 1), $addedPoints, isset($conf[$pageNo . '_after_reply']) ? $conf[$pageNo . '_after_reply'] : ''),
            $output
        );
        outputHtml($output);
        fwrite($teamHandle, 'CORRECT page: ' . $pageNo . ' attempt:' . $attempt . ' correct_answer:' . $correctAnswer . ' . answer:' . $answer . ' elapsed:' . $elapsedTime . "\n");
        fwrite($teamHandle, 'points:' . $currentPoints . "\n");
        fwrite($teamHandle, 'timestamp page: ' . $pageNo . ' timestamp now:' . time() . "\n");
        fclose($teamHandle);
        exit;
    }

    fwrite($teamHandle, 'WRONG page: ' . $pageNo . ' attempt:' . $attempt . ' correct_answer:' . $correctAnswer . ' . answer:' . $answer . ' elapsed:' . $elapsedTime . "\n");
    $attempt++;

    if ($attempt > 3) {
        $output = translateHtml(readFileContents($path . 'html' . DIRECTORY_SEPARATOR . 'after_many_mistakes.htm'));
        $output = str_replace(
            array('{page}', '{after_reply}', '{answer}'),
            array(hiddenPageNum((int) $pageNo + 1), isset($conf[$pageNo . '_after_reply']) ? $conf[$pageNo . '_after_reply'] : '', $correctAnswer),
            $output
        );
        outputHtml($output);
        fwrite($teamHandle, 'MANY_WRONG page: ' . $pageNo . ' attempt:' . $attempt . ' correct_answer:' . $correctAnswer . ' . answer:' . $answer . ' elapsed:' . $elapsedTime . "\n");
        fwrite($teamHandle, 'points:' . $currentPoints . "\n");
        fwrite($teamHandle, 'timestamp page: ' . $pageNo . ' timestamp now:' . time() . "\n");
        fclose($teamHandle);
        exit;
    }

    if (isset($conf[$pageNo . '_q_img']) && $conf[$pageNo . '_q_img'] !== '') {
        $question .= '<br/><img style="width:100%; max-width:800px;" src="' . $conf[$pageNo . '_q_img'] . '"./>';
    }
    $output = translateHtml(readFileContents($path . 'html' . DIRECTORY_SEPARATOR . 'wrong.htm'));
    $output = str_replace(
        array('{page}', '{attempt}', '{question}'),
        array($pageParam, $attempt, $question),
        $output
    );
    outputHtml($output);
    fclose($teamHandle);
    exit;
}

fclose($teamHandle);
$logFile = $path . 'logs' . DIRECTORY_SEPARATOR . 'main.log';
$logDirectory = dirname($logFile);
if (!is_dir($logDirectory)) {
    mkdir($logDirectory, 0775, true);
}
file_put_contents($logFile, 'ooo ' . $now . "\n\n", FILE_APPEND);
echo "error...\n";

function loadConfig($file)
{
    if (!file_exists($file)) {
        return array();
    }

    $text = readFileContents($file);
    preg_match_all('/^([^:\r\n]+):\r?\n(.*?)(?=^[^:\r\n]+:\r?\n|\z)/ms', $text, $matches, PREG_SET_ORDER);
    $config = array();
    foreach ($matches as $match) {
        $config[$match[1]] = rtrim($match[2], "\r\n");
    }
    return $config;
}

function translateHtml($html)
{
    global $conf, $game, $team, $currentPoints, $elapsedTime;
    $values = $conf;
    $values['game'] = $game;
    $values['team'] = $team;
    $values['current_points'] = $currentPoints;
    $values['elapsed_time'] = $elapsedTime;
    foreach ($values as $key => $value) {
        $html = str_replace('{' . $key . '}', $value, $html);
    }
    return str_replace('treasure.pl', 'treasure.php', $html);
}

function outputHtml($html)
{
    echo $html;
}

function hiddenPageNum($number)
{
    return '7' . sprintf('%02d', $number) . '3432';
}

function calcElapsed($teamText, $config)
{
    $startTime = time() - 1;
    if (preg_match('/start timestamp:(\d*)/s', $teamText, $matches)) {
        $startTime = (int) $matches[1];
    }

    $timeToReduce = 0;
    if (isset($config['possible_pause']) && $config['possible_pause'] !== '') {
        $pausePage = $config['possible_pause'];
        $afterPausePage = (int) $pausePage + 1;
        $beforeTimestamp = 0;
        $afterTimestamp = 0;
        if (preg_match('/.*timestamp page: ' . preg_quote($pausePage, '/') . ' timestamp now:(\d*)/s', $teamText, $matches)) {
            $beforeTimestamp = (int) $matches[1];
        }
        if (preg_match('/.*timestamp page: ' . $afterPausePage . ' timestamp now:(\d*)/s', $teamText, $matches)) {
            $afterTimestamp = (int) $matches[1];
        }
        if ($beforeTimestamp > 0 && $afterTimestamp > 0) {
            $timeToReduce = $afterTimestamp - $beforeTimestamp;
        }
    }

    if (preg_match('/.*timestamp page: (\d*) timestamp now:(\d*)/s', $teamText, $matches)) {
        return wdhms((int) $matches[2] - $startTime - $timeToReduce);
    }
    return 0;
}

function wdhms($seconds)
{
    $sign = $seconds < 0 ? '-' : '';
    $seconds = abs($seconds);
    $minutes = (int) floor($seconds / 60);
    $seconds %= 60;
    $hours = (int) floor($minutes / 60);
    $minutes %= 60;
    $days = (int) floor($hours / 24);
    $hours %= 24;
    $weeks = (int) floor($days / 7);
    $days %= 7;

    $result = sprintf('%02d', $seconds);
    if ($minutes || $hours || $days || $weeks) {
        $result = sprintf('%02d:', $minutes) . $result;
    }
    if ($hours || $days || $weeks) {
        $result = sprintf('%02d:', $hours) . $result;
    }
    if ($days > 0) {
        $result = $days . 'd ' . $result;
        if ($weeks > 0) {
            $result = $weeks . 'w' . $result;
        }
    }
    return $sign . $result;
}

function readFileContents($file)
{
    $contents = @file_get_contents($file);
    if ($contents === false) {
        http_response_code(404);
        exit('Cannot read ' . $file);
    }
    return $contents;
}

function showAdmin($path, $game, $config)
{
    echo '<table border=1>';
    foreach (glob($path . 'data' . DIRECTORY_SEPARATOR . 'team*') as $file) {
        $theTeam = preg_replace('/.*_/', '', $file);
        $theTeam = preg_replace('/\.data$/', '', $theTeam);
        $teamText = readFileContents($file);
        echo '<tr><td><a href="treasure.php?game=' . $game . '&team=' . $theTeam . '&showMap=true">' . $theTeam . '</a></td>';
        echo '<td><a href="treasure.php?game=' . $game . '&team=' . $theTeam . '&showLog=true">log</a></td>';
        echo '<td>elapsed until last timestamp:' . calcElapsed($teamText, $config) . '</td></tr>';
    }
    echo '</table>';
}

function showMap($file, $path)
{
    $teamText = readFileContents($file);
    $values = '';
    if (preg_match_all('/GEO: lat (.*?) lon (.*?)\n/', $teamText, $matches, PREG_SET_ORDER)) {
        foreach ($matches as $match) {
            $values .= ' [' . $match[1] . ', ' . $match[2] . ', ""],';
        }
    }
    echo str_replace('{values}', $values, readFileContents($path . 'html' . DIRECTORY_SEPARATOR . 'showMap.htm'));
    exit;
}

function showLog($file)
{
    echo $file . " <br/>\n <pre>" . readFileContents($file) . "</pre>";
    echo "___________________<br/><br/><br/><br/>\n";
    exit;
}
